diff --git a/notxv6/ph.c b/notxv6/ph.c
index 6df1500..7d08e31 100644
--- a/notxv6/ph.c
+++ b/notxv6/ph.c
@@ -16,6 +16,7 @@ struct entry {
 struct entry *table[NBUCKET];
 int keys[NKEYS];
 int nthread = 1;
+pthread_mutex_t lock[NBUCKET];
 
 double
 now()
@@ -46,6 +47,7 @@ void put(int key, int value)
     if (e->key == key)
       break;
   }
+  pthread_mutex_lock(&lock[i]); 
   if(e){
     // update the existing key.
     e->value = value;
@@ -53,6 +55,7 @@ void put(int key, int value)
     // the new is new.
     insert(key, value, &table[i], table[i]);
   }
+  pthread_mutex_unlock(&lock[i]);
 }
 
 static struct entry*
@@ -76,7 +79,9 @@ put_thread(void *xa)
   int b = NKEYS/nthread;
 
   for (int i = 0; i < b; i++) {
+    //pthread_mutex_lock(&lock);       // acquire lock
     put(keys[b*n + i], n);
+    //pthread_mutex_unlock(&lock);     // release lock
   }
 
   return NULL;
@@ -115,6 +120,10 @@ main(int argc, char *argv[])
     keys[i] = random();
   }
 
+  // init lock
+  for (int i = 0; i < NBUCKET; i++) {
+    pthread_mutex_init(&lock[i], NULL);
+  }
   //
   // first the puts
   //
